### PR TITLE
Fixed #31392 -- Avoided unnecessary SchemaEditor.effective_default() calls when altering a field.

### PR DESCRIPTION
**Pull request by fengsp**

### CODE REVIEW

File: django/db/backends/base/schema.py [hunk 1]
```diff
@@ -675,17 +675,13 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
...
-            old_default != new_default and
-            new_default is not None and
-            not self.skip_default(new_field)
-        )
-        if needs_database_default:
-            actions.append(self._alter_column_default_sql(model, old_field, new_field))
+        needs_database_default = False
+        if old_field.null and not new_field.null and not self.skip_default(new_field):
+            old_default = self.effective_default(old_field)
+            new_default = self.effective_default(new_field)
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: We should avoid calling `self.effective_default(new_field)` twice (here and in [line 721](https://github.com/django/django/pull/12612/files#diff-b69190ab88f6c5737b2562d94d7bf36bR721)). Maybe we can make it lazy or refactor this change a bit.

Comment (fengsp) [reply django/db/backends/base/schema.py]: I noticed this, and `_alter_column_default_sql` is also calling `self.effective_default(new_field)`.

How about:
```
needs_database_default = False
four_way_default_alteration = False
if old_field.null and not new_field.null:
    old_default = self.effective_default(old_field)
    new_default = self.effective_default(new_field)
    if old_default != new_default and new_default is not None and not self.skip_default(new_field):
        # Default change?
        needs_database_default = True
        actions.append(self._alter_column_default_sql(model, old_field, new_field))
    if new_field.has_default():
        # Only if we have a default and there is a change from NULL to NOT NULL
        four_way_default_alteration = True
```

I don't like this though, it is less obvious where `new_default` comes from.

Comment (felixxm) [reply django/db/backends/base/schema.py]: We use `new_default` only when `old_field.null and not new_field.null` so IMO it's fine to use
```diff
diff --git a/django/db/backends/base/schema.py b/django/db/backends/base/schema.py
index bfccf5e8fb..8cd5e11bbf 100644
--- a/django/db/backends/base/schema.py
+++ b/django/db/backends/base/schema.py
@@ -675,17 +675,17 @@ class BaseDatabaseSchemaEditor:
         #  3. Replace NULL constraint with NOT NULL
         #  4. Drop the default again.
         # Default change?
-        old_default = self.effective_default(old_field)
-        new_default = self.effective_default(new_field)
-        needs_database_default = (
-            old_field.null and
-            not new_field.null and
-            old_default != new_default and
-            new_default is not None and
-            not self.skip_default(new_field)
-        )
-        if needs_database_default:
-            actions.append(self._alter_column_default_sql(model, old_field, new_field))
+        needs_database_default = False
+        if old_field.null and not new_field.null:
+            old_default = self.effective_default(old_field)
+            new_default = self.effective_default(new_field)
+            if (
+                not self.skip_default(new_field) and
+                old_default != new_default and
+                new_default is not None
+            ):
+                needs_database_default = True
+                actions.append(self._alter_column_default_sql(model, old_field, new_field))
         # Nullability change?
         if old_field.null != new_field.null:
             fragment = self._alter_column_null_sql(model, old_field, new_field)

```

Comment (fengsp) [reply django/db/backends/base/schema.py]: Thanks for the advice, I'll update it.

**Pull request merged**